upf/dnf.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import upf.environment
+import upf.operators as op
+import upf.walkers as walkers
+from upf.exceptions import UPFUnraisableError
+from upf.walkers.dag import DagWalker
+from upf.fnode import FNode
+from typing import List, Tuple
+from itertools import product
+
+
+class Nnf():
+    """Class used to transform a logic expression into the equivalent
+    Negational Normal Form expression.
+
+    This is done first by removing all the Implications and Equalities,
+    then by pushing all the not to the leaves of the Tree representing the expression."""
+
+    def __init__(self, env: 'upf.environment.Environment'):
+        self.env = env
+        self.manager = env.expression_manager
+
+    def get_nnf_expression(self, expression: FNode) -> FNode:
+        """Function used to transform a logic expression into the equivalent
+        Negational Normal Form expression.
+
+        This is done first by removing all the Implications and Equalities,
+        and by pushing all the not to the leaves of the Tree representing the expression.
+
+        For example, the form: !(a => (b && c)) becomes:
+        a && (!b || !c), therefore it is a NNF."""
+        stack: List[Tuple[bool, FNode, bool]] = []
+        stack.append((True, expression, False))
+        solved: List[FNode] = []
+        while len(stack) > 0:
+            p, e, status = stack.pop()
+            if status:
+                if e.is_and():
+                    args = [solved.pop() for i in range(len(e.args()))]
```

Comment:
```suggestion
                    args = [solved.pop() for _ in range(len(e.args()))]
```

Reply:
@Framba-Luca this is very minor, but is useful to avoid nasty bugs when you use variable `i` after this line without re-initializing it (unfortunately it is valid in python). Instead `_` marks an unneeded value and cannot be reused

Reply:
Just for information, there is even another way to do the same thing, but probably the repeated `pop()` is much clearer:
```suggestion
                    arity = len(e.args())
                    args = list(reversed(solved[-arity:]))
                    del solved[-arity:]
```